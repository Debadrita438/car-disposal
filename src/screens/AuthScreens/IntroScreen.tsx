import React, {useRef, useState} from 'react';
import {
  Animated,
  Dimensions,
  ScrollView,
  StyleSheet,
  View,
  Text,
  Image,
  TouchableOpacity,
} from 'react-native';
import {
  CheckIcon,
  FirstIntro,
  FourthIntro,
  RightArrow,
  SecondIntro,
  ThirdIntro,
} from '../../assets';
import {Colors, Fonts} from '../../theme';

const {width, height} = Dimensions.get('window');
const data = [
  {
    title: 'Schedule Vehicle Pick Up',
    description:
      'I would like to dispose of my car and need information on the process. Please provide details on the available options.',
    image: (
      <FirstIntro
        height={height / 1.5}
        width={width}
        style={{marginTop: -30}}
      />
    ),
  },
  {
    title: 'Vehicle Details',
    description:
      'Please share the key details of your vehicle. We need the Vehicle Identification Number (VIN), Year, Make, and Model to ensure everything is processed correctly and responsibly.',
    image: (
      <Image
        source={require('../../assets/images/intro2.png')}
        style={{width: '100%', height: '66%'}}
        resizeMode="contain"
      />
    ),
  },
  {
    title: 'Your Car will be Picked up in \n the Schedule Time',
    description:
      "Your car will be picked up at the scheduled time.\n Please ensure it's ready for Pick Up.",
    image: (
      <Image
        source={require('../../assets/images/intro3.png')}
        style={{width: '100%', height: '66%'}}
        resizeMode="contain"
      />
    ),
  },
  {
    title: 'Vehicle Has Been Picked Up',
    description:
      'Your car has been successfully Picked Up and Transported to a Secured Facility',
    image: (
      <FourthIntro
        height={height / 1.5}
        width={width}
        style={{marginTop: -30}}
      />
    ),
  },
];

const IntroScreen = ({navigation}) => {
  const scrollValue = useRef(new Animated.Value(0)).current;
  const scrollViewRef = useRef(null);
  const [currentIndex, setCurrentIndex] = useState(0);

  const translateX = scrollValue.interpolate({
    inputRange: [0, width],
    outputRange: [0, 6.7 * (data.length - 1)],
  });

  const handleArrowPress = () => {
    if (currentIndex < data.length - 1) {
      const nextIndex = currentIndex + 1;
      setCurrentIndex(nextIndex);
      scrollViewRef?.current?.scrollTo({
        x: nextIndex * width,
        animated: true,
      });
    } else {
      // Navigate to the next screen when on the last index
      navigation.navigate('LoginScreen'); // Replace with your target screen
    }
  };

  const handleScroll = event => {
    const offsetX = event.nativeEvent.contentOffset.x;
    const index = Math.floor(offsetX / width);
    setCurrentIndex(index);
  };

  return (
    <View style={styles.container}>
      <ScrollView
        horizontal
        pagingEnabled
        decelerationRate="fast"
        showsHorizontalScrollIndicator={false}
        ref={scrollViewRef}
        onScroll={Animated.event(
          [{nativeEvent: {contentOffset: {x: scrollValue}}}],
          {useNativeDriver: false, listener: handleScroll},
        )}>
        {data.map((item, index) => (
          <View style={styles.card} key={index}>
            {item.image}
            <Text style={Fonts.infoHeaderText}>{item.title}</Text>
            <Text style={Fonts.infoDescText}>{item.description}</Text>
          </View>
        ))}
      </ScrollView>
      <View style={styles.indicatorContainer} pointerEvents="none">
        {data.map((_, index) => (
          <View style={styles.indicator} key={index} />
        ))}
        <Animated.View
          style={[
            styles.activeIndicator,
            {position: 'absolute', transform: [{translateX}]},
          ]}
        />
      </View>
      <View
        style={[
          styles.bottomContainer,
          {
            justifyContent:
              currentIndex < data.length - 1 ? 'space-between' : 'flex-end',
          },
        ]}>
        {currentIndex < data.length - 1 && (
          <TouchableOpacity onPress={() => navigation.navigate('LoginScreen')}>
            <Text style={styles.skipText}>Skip</Text>
          </TouchableOpacity>
        )}
        <TouchableOpacity
          style={[
            styles.arrowButton,
            currentIndex === data.length - 1 && styles.leftArrow,
          ]}
          onPress={handleArrowPress}>
          {currentIndex < data.length - 1 ? <RightArrow /> : <CheckIcon />}
        </TouchableOpacity>
      </View>
    </View>
  );
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: Colors.whiteColor,
  },
  card: {
    width,
    alignItems: 'center',
    padding: 20,
  },
  image: {
    width: 250,
    height: 250,
    resizeMode: 'contain',
  },
  title: {
    fontSize: 22,
    fontWeight: 'bold',
    color: Colors.textColor,
    marginTop: 20,
    textAlign: 'center',
  },
  description: {
    fontSize: 13,
    color: Colors.textColor,
    marginTop: 10,
    textAlign: 'center',
  },
  indicatorContainer: {
    flexDirection: 'row',
    alignSelf: 'center',
    marginTop: 20,
    position: 'absolute',
    bottom: 50,
  },
  indicator: {
    height: 10,
    width: 10,
    borderRadius: 5,
    backgroundColor: Colors.inactiveDot,
    marginHorizontal: 5,
  },
  activeIndicator: {
    height: 10,
    width: 10,
    borderRadius: 5,
    backgroundColor: Colors.baseColor,
    marginHorizontal: 5,
  },
  bottomContainer: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    alignItems: 'center',
    paddingHorizontal: 20,
    position: 'absolute',
    bottom: 30,
    width: '100%',
  },
  skipText: {
    fontSize: 16,
    color: Colors.textColor,
    fontFamily: Fonts.regular,
  },
  arrowButton: {
    backgroundColor: Colors.baseColor,
    width: 50,
    height: 50,
    borderRadius: 25,
    justifyContent: 'center',
    alignItems: 'center',
  },
  leftArrow: {
    alignSelf: 'flex-start', // Align to the left when on the last index
  },
});

export default IntroScreen;
