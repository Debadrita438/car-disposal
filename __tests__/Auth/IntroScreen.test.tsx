import React from 'react';
import {fireEvent, render} from '@testing-library/react-native';
import renderer from 'react-test-renderer';
import IntroScreen from '../../src/screens/AuthScreens/IntroScreen';
import {TouchableOpacity} from 'react-native';

jest.mock('@react-navigation/native', () => ({
  useNavigation: () => ({
    navigate: jest.fn(),
  }),
}));

describe('IntroScreen', () => {
  const mockNavigation = {navigate: jest.fn()}; // Mock navigation object

  beforeEach(() => {
    jest.clearAllMocks(); // Reset mocks before each test
  });

  it('renders all intro steps correctly', () => {
    const {getByLabelText, getByText} = render(
      <IntroScreen navigation={mockNavigation} />,
    );

    // Verify step content
    expect(getByLabelText('FirstIntroMock')).toBeTruthy();
    expect(getByText('Schedule Vehicle Pick Up')).toBeTruthy();

    expect(getByLabelText('FourthIntroMock')).toBeTruthy();
    expect(getByText('Vehicle Has Been Picked Up')).toBeTruthy();
  });

  it('navigates to LoginScreen when Skip is pressed', () => {
    const {getByText} = render(<IntroScreen navigation={mockNavigation} />);

    const skipButton = getByText('Skip');
    fireEvent.press(skipButton);

    expect(mockNavigation.navigate).toHaveBeenCalledWith('LoginScreen');
  });

  it('renders RightArrow initially', () => {
    const {getByLabelText} = render(
      <IntroScreen navigation={mockNavigation} />,
    );
    const rightIcon = getByLabelText('right');
    expect(rightIcon).toBeTruthy();
  });

  it('renders CheckIcon when on the last index', () => {
    const {getByLabelText, getByTestId} = render(
      <IntroScreen navigation={mockNavigation} />,
    );
    const nextButton = getByTestId('next');

    // Simulate presses to navigate to the last item
    fireEvent.press(nextButton);
    fireEvent.press(nextButton);
    fireEvent.press(nextButton);

    const checkIcon = getByLabelText('check');
    expect(checkIcon).toBeTruthy();
  });

  it('triggers handleArrowPress on press', () => {
    const mockHandleArrowPress = jest.fn();
    const {getByTestId} = render(
      <TouchableOpacity onPress={mockHandleArrowPress} testID="next" />,
    );

    const nextButton = getByTestId('next');
    fireEvent.press(nextButton);

    expect(mockHandleArrowPress).toHaveBeenCalled();
  });
});

test('renders correctly', () => {
  const mockNavigation = {navigate: jest.fn()};
  const tree = renderer
    .create(<IntroScreen navigation={mockNavigation} />)
    .toJSON();
  expect(tree).toMatchSnapshot();
});
