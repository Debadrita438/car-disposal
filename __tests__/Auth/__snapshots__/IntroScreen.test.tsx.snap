// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    {
      "backgroundColor": "#ffffff",
      "flex": 1,
    }
  }
>
  <RCTScrollView
    decelerationRate="fast"
    horizontal={true}
    onScroll={[Function]}
    pagingEnabled={true}
    showsHorizontalScrollIndicator={false}
  >
    <View>
      <View
        style={
          {
            "alignItems": "center",
            "padding": 20,
            "width": 750,
          }
        }
      >
        <SvgComponent
          accessibilityLabel="FirstIntroMock"
          height={889.3333333333334}
          style={
            {
              "marginTop": -30,
            }
          }
          width={750}
        />
        <Text
          style={
            {
              "color": "#3D459B",
              "fontFamily": "Poppins-SemiBold",
              "fontSize": 22,
              "textAlign": "center",
            }
          }
        >
          Schedule Vehicle Pick Up
        </Text>
        <Text
          style={
            {
              "color": "#3D459B",
              "fontFamily": "Poppins-Regular",
              "fontSize": 13,
              "textAlign": "center",
            }
          }
        >
          I would like to dispose of my car and need information on the process. Please provide details on the available options.
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "padding": 20,
            "width": 750,
          }
        }
      >
        <Image
          resizeMode="contain"
          source={
            {
              "testUri": "../../../src/assets/images/intro2.png",
            }
          }
          style={
            {
              "height": "66%",
              "width": "100%",
            }
          }
        />
        <Text
          style={
            {
              "color": "#3D459B",
              "fontFamily": "Poppins-SemiBold",
              "fontSize": 22,
              "textAlign": "center",
            }
          }
        >
          Vehicle Details
        </Text>
        <Text
          style={
            {
              "color": "#3D459B",
              "fontFamily": "Poppins-Regular",
              "fontSize": 13,
              "textAlign": "center",
            }
          }
        >
          Please share the key details of your vehicle. We need the Vehicle Identification Number (VIN), Year, Make, and Model to ensure everything is processed correctly and responsibly.
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "padding": 20,
            "width": 750,
          }
        }
      >
        <Image
          resizeMode="contain"
          source={
            {
              "testUri": "../../../src/assets/images/intro3.png",
            }
          }
          style={
            {
              "height": "66%",
              "width": "100%",
            }
          }
        />
        <Text
          style={
            {
              "color": "#3D459B",
              "fontFamily": "Poppins-SemiBold",
              "fontSize": 22,
              "textAlign": "center",
            }
          }
        >
          Your Car will be Picked up in 
 the Schedule Time
        </Text>
        <Text
          style={
            {
              "color": "#3D459B",
              "fontFamily": "Poppins-Regular",
              "fontSize": 13,
              "textAlign": "center",
            }
          }
        >
          Your car will be picked up at the scheduled time.
 Please ensure it's ready for Pick Up.
        </Text>
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "padding": 20,
            "width": 750,
          }
        }
      >
        <SvgComponent
          accessibilityLabel="FourthIntroMock"
          height={889.3333333333334}
          style={
            {
              "marginTop": -30,
            }
          }
          width={750}
        />
        <Text
          style={
            {
              "color": "#3D459B",
              "fontFamily": "Poppins-SemiBold",
              "fontSize": 22,
              "textAlign": "center",
            }
          }
        >
          Vehicle Has Been Picked Up
        </Text>
        <Text
          style={
            {
              "color": "#3D459B",
              "fontFamily": "Poppins-Regular",
              "fontSize": 13,
              "textAlign": "center",
            }
          }
        >
          Your car has been successfully Picked Up and Transported to a Secured Facility
        </Text>
      </View>
    </View>
  </RCTScrollView>
  <View
    pointerEvents="none"
    style={
      {
        "alignSelf": "center",
        "bottom": 50,
        "flexDirection": "row",
        "marginTop": 20,
        "position": "absolute",
      }
    }
  >
    <View
      style={
        {
          "backgroundColor": "#6C7E8E",
          "borderRadius": 5,
          "height": 10,
          "marginHorizontal": 5,
          "width": 10,
        }
      }
    />
    <View
      style={
        {
          "backgroundColor": "#6C7E8E",
          "borderRadius": 5,
          "height": 10,
          "marginHorizontal": 5,
          "width": 10,
        }
      }
    />
    <View
      style={
        {
          "backgroundColor": "#6C7E8E",
          "borderRadius": 5,
          "height": 10,
          "marginHorizontal": 5,
          "width": 10,
        }
      }
    />
    <View
      style={
        {
          "backgroundColor": "#6C7E8E",
          "borderRadius": 5,
          "height": 10,
          "marginHorizontal": 5,
          "width": 10,
        }
      }
    />
    <View
      collapsable={false}
      style={
        {
          "backgroundColor": "#333C99",
          "borderRadius": 5,
          "height": 10,
          "marginHorizontal": 5,
          "position": "absolute",
          "transform": [
            {
              "translateX": 0,
            },
          ],
          "width": 10,
        }
      }
    />
  </View>
  <View
    style={
      [
        {
          "alignItems": "center",
          "bottom": 30,
          "flexDirection": "row",
          "justifyContent": "space-between",
          "paddingHorizontal": 20,
          "position": "absolute",
          "width": "100%",
        },
        {
          "justifyContent": "space-between",
        },
      ]
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <Text
        style={
          {
            "color": "#3D459B",
            "fontFamily": "Poppins-Regular",
            "fontSize": 16,
          }
        }
      >
        Skip
      </Text>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "backgroundColor": "#333C99",
          "borderRadius": 25,
          "height": 50,
          "justifyContent": "center",
          "opacity": 1,
          "width": 50,
        }
      }
      testID="next"
    >
      <SvgComponent
        accessibilityLabel="right"
      />
    </View>
  </View>
</View>
`;
